sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala

```diff
@@ -324,6 +325,124 @@ object LogicalPlanIntegrity {
+        val danglingReferences = n.references.filter {
+          a => a.resolved && !inputExprIds.contains(a.exprId)
+        }.map(_.qualifiedName)
+        if (danglingReferences.nonEmpty) {
+          Some(s"Aliases ${danglingReferences.mkString(", ")} are dangling " +
+            s"in the references for plan:\n ${n.treeString}")
+        } else {
+          None
+        }
+    }.flatten
+  }
+
+  /**
+   * Validate that the grouping key types in Aggregate plans are valid.
+   * Returns an error message if the check fails, or None if it succeeds.
+   */
+  def validateGroupByTypes(plan: LogicalPlan): Option[String] = {
+    plan.collectFirst {
+      case a @ Aggregate(groupingExprs, _, _) =>
+        val badExprs = groupingExprs.filter(_.dataType.isInstanceOf[MapType]).map(_.toString)
+        if (badExprs.nonEmpty) {
+          Some(s"Grouping expressions ${badExprs.mkString(", ")} cannot be of type Map " +
+            s"for plan:\n ${a.treeString}")
+        } else {
+          None
+        }
+    }.flatten
+  }
+
+  /**
+   * Validate that the aggregation expressions in Aggregate plans are valid.
+   * Returns an error message if the check fails, or None if it succeeds.
+   */
+  def validateAggregateExpressions(plan: LogicalPlan): Option[String] = {
+    /**
+    * Returns true if all the exprIds, referenced in an Expression <subExpr>
+    * (used in the aggregate expressions of an Aggregate),
+    * are either used as measures (under an AggregateFunction) or are explicitly grouped,
+    * given the <groupingExprIds> of all grouping expressions
+    * (used in the grouping expressions of the same Aggregate)
+
+    * I.e:
+    * <groupingExprIds> can be used "freely" in an AggregateExpressions, all other exprIds can only
+    * be used under an AggregateFunction (i.e. aggregated).
+     */
+    def isAggregateSubExpressionValid(
+          groupingExpressions: Seq[Expression],
+          groupingExprIds: Set[ExprId],
+          subExpr: Expression): Boolean = {
+      // First collect all non-grouping ExprIds in <subExpr>.
+      val restrictedExprIds = subExpr.collect {
+        case g: Attribute if !groupingExprIds.contains(g.exprId) => g.exprId
+      }.groupBy(identity).mapValues(_.size)
+
+      // Second collect all non-grouping ExprIds under an AggregateFunction in <subExpr>.
+      val restrictedExprIdsUnderAggregateFunction = subExpr.flatMap {
+        case a: AggregateExpression => a.collect {
+          case g: Attribute if !groupingExprIds.contains(g.exprId) => g.exprId
+        }
+        case _ => Seq.empty
+      }.groupBy(identity).mapValues(_.size)
+
+      // Validate a) that all non-grouping exprIds are used under some aggregate function,
+      // i.e. all such exprIds are aggregated (measures) OR
+      // b) a grouping expression can be used in sub-expressions
+      // for aggregate expressions of an Aggregate.
+      restrictedExprIds == restrictedExprIdsUnderAggregateFunction ||
```

Comment:
I'm wondering if this shortcut really helps. For a simple query like `select a ... group by a, b`, we build two maps with `.groupBy` to know that `select a ...` is valid. Without the shortcut, it's simply ` Seq(attr_a, attr_b).exists(_.semanticEquals(attr_a))` which should be much faster than building two maps and compare.

Reply:
the validation needs to check the exprids; we've seen quite a few cases where rewrites mess those up.

Reply:
> the validation needs to check the exprids; we've seen quite a few cases where rewrites mess those up.

It will be great to elaborate on this a bit more. I thought this is only for performance, does it affect correctness?

Reply:
synced offline